### PR TITLE
nes_vt: APU improvements

### PR DESCRIPTION
**Pull request by daveshah1**
- Add support for the VT03+ second APU channel (with all functionality except DWS/PCM)
 - Add support for the VT03+ PCM mode
 - Add support for the VT3x dual-channel PCM mode

### CODE REVIEW

**Comment by DavidHaywood** [comment]
thank you for the continued interest in these systems :-)

File: src/devices/sound/nes_vt_apu.cpp [hunk 1]
```diff
@@ -0,0 +1,289 @@
+// license:BSD-3-Clause
+// copyright-holders:David Shah
+/*****************************************************************************
+
+  MAME/MESS VTxx APU CORE
+
+ *****************************************************************************/
+
+#include "nes_vt_apu.h"
+#include "emu.h"
```
Comment (cuavas) [inline src/devices/sound/nes_vt_apu.cpp]: This won't build properly with precompiled headers disabled - `emu.h` must be the first thing in each compilation unit.

File: src/devices/sound/nes_vt_apu.cpp [hunk 2]
```diff
@@ -0,0 +1,289 @@
...
+	: nesapu_device(mconfig, tag, NES_VT_APU, owner, clock),
+	  m_xop2(*this, "nesapu_vt_slave"),
+	  m_rom_read_cb(*this)
+{
+
+}
+
+void nesapu_vt_device::device_start()
+{
+	apu_init();
```
Comment (cuavas) [inline src/devices/sound/nes_vt_apu.cpp]: If you can't call the base class `device_start` it's usually a sign that the design is broken.  If there's stuff the regular NES APU does that the VT APU doesn't, that should be split into a derived class, so both APUs can use a common base.

File: src/devices/sound/nes_vt_apu.cpp [hunk 3]
```diff
@@ -0,0 +1,289 @@
...
+	std::unique_ptr<stream_sample_t[]> sbuf = std::make_unique<stream_sample_t[]>(samples);
+
+	stream_sample_t *pptr = pbuf.get(), *sptr = sbuf.get();
+
+	// Dual legacy sound generators
+	nesapu_device::sound_stream_update(stream, inputs, &pptr, samples);
+	m_xop2->sound_stream_update(stream, inputs, &sptr, samples);
+
+	int accum;
+	memset( outputs[0], 0, samples*sizeof(*outputs[0]) );
```
Comment (cuavas) [inline src/devices/sound/nes_vt_apu.cpp]: This function isn't consistent in formatting within itself - there are extra spaces inside the parentheses for the `memset` call that aren't used on any of the other function calls here.

File: src/devices/sound/nes_vt_apu.cpp [hunk 4]
```diff
@@ -0,0 +1,289 @@
...
+		else if (accum < -128)
+			accum = -128;
+
+		*(outputs[0]++)=accum<<8;
+
+	}
+}
+
+s8 nesapu_vt_device::vt03_pcm(apu_vt_t::vt03_pcm_t *ch) {
+	if (ch->enabled) {
```
Comment (cuavas) [inline src/devices/sound/nes_vt_apu.cpp]: Why are the opening braces for these two blocks at the end of a line, when all the others are on their own lines?

**Comment by cuavas** [comment]
I'm sorry, but this needs to be reverted.  The implementation is sloppy and it completely breaks previously-working systems.  The freeze for this month's release is rapidly approaching, and the issues have not been acknowledged.  Please address the issues and resubmit.

**Comment by Tafoid** [comment]
Comments left on the commit to master mention some breakage:  https://github.com/mamedev/mame/commit/6d3238ead8c25c36629bf5654ba0b821cf346ea4

**Pull request merged**